File: mycode/server/routers/sessions.py
"""Session management API endpoints."""

from __future__ import annotations

import os

from fastapi import APIRouter, HTTPException

from mycode.core.config import get_settings, resolve_provider
from mycode.server.deps import RunManagerDep, StoreDep
from mycode.server.schemas import SessionCreateRequest

router = APIRouter(prefix="/sessions", tags=["sessions"])


@router.post("")
async def create_session(req: SessionCreateRequest, store: StoreDep):
    cwd = os.path.abspath(req.cwd or os.getcwd())
    settings = get_settings(cwd)
    resolved = resolve_provider(settings, provider_name=req.provider, model=req.model, api_base=req.api_base)
    return await store.create_session(
        req.title,
        provider=resolved.provider,
        model=resolved.model,
        cwd=cwd,
        api_base=resolved.api_base,
    )


@router.get("")
async def list_sessions(store: StoreDep, runs: RunManagerDep, cwd: str | None = None):
    sessions = await store.list_sessions(cwd=cwd)
    for session in sessions:
        session["is_running"] = await runs.has_active_run(session.get("id", ""))
    return {"sessions": sessions}


@router.get("/{session_id}")
async def load_session(session_id: str, store: StoreDep, runs: RunManagerDep):
    active = await runs.snapshot_session(session_id)
    if active:
        data = await store.load_session(session_id)
        session = data.get("session") if data else None
        return {
            "session": session,
            "messages": active["messages"],
            "active_run": active["run"],
            "pending_events": active["pending_events"],
        }

    data = await store.load_session(session_id)
    if not data:
        return {"session": None, "messages": [], "active_run": None, "pending_events": []}
    return {**data, "active_run": None, "pending_events": []}


@router.delete("/{session_id}")
async def delete_session(session_id: str, store: StoreDep, runs: RunManagerDep):
    if await runs.has_active_run(session_id):
        raise HTTPException(status_code=409, detail="session has a running task")
    await store.delete_session(session_id)
    return {"status": "ok"}


@router.post("/{session_id}/clear")
async def clear_session(session_id: str, store: StoreDep, runs: RunManagerDep):
    if await runs.has_active_run(session_id):
        raise HTTPException(status_code=409, detail="session has a running task")
    await store.clear_session(session_id)
    return {"status": "ok"}

File: mycode-go/internal/server/sessions.go
package server

import (
	"net/http"
	"strings"

	"github.com/legibet/mycode-go/internal/config"
	"github.com/legibet/mycode-go/internal/message"
	"github.com/legibet/mycode-go/internal/session"
)

type sessionCreateRequest struct {
	Title    string `json:"title"`
	Provider string `json:"provider"`
	Model    string `json:"model"`
	CWD      string `json:"cwd"`
	APIBase  string `json:"api_base"`
}

type sessionResponse struct {
	Session       any               `json:"session"`
	Messages      []message.Message `json:"messages"`
	ActiveRun     any               `json:"active_run"`
	PendingEvents []map[string]any  `json:"pending_events"`
}

type sessionListItem struct {
	session.Meta
	IsRunning bool `json:"is_running"`
}

func (a *app) handleCreateSession(w http.ResponseWriter, r *http.Request) {
	var req sessionCreateRequest
	if err := decodeJSON(r, &req); err != nil {
		writeDetailError(w, http.StatusBadRequest, err.Error())
		return
	}

	cwd := requestCWD(req.CWD)
	settings, err := config.Load(cwd)
	if err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}
	resolved, err := config.ResolveProvider(settings, req.Provider, req.Model, "", req.APIBase, "")
	if err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}

	data, err := a.store.CreateSession(req.Title, "", resolved.ProviderType, resolved.Model, cwd, resolved.APIBase)
	if err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, data)
}

func (a *app) handleListSessions(w http.ResponseWriter, r *http.Request) {
	cwd := strings.TrimSpace(r.URL.Query().Get("cwd"))
	items, err := a.store.ListSessions(cwd)
	if err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}

	sessions := make([]sessionListItem, 0, len(items))
	for _, item := range items {
		sessions = append(sessions, sessionListItem{
			Meta:      item,
			IsRunning: a.runs.hasActiveRun(item.ID),
		})
	}
	writeJSON(w, http.StatusOK, map[string]any{"sessions": sessions})
}

func (a *app) handleLoadSession(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("session_id")
	data, err := a.store.LoadSession(sessionID)
	if err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}

	resp := sessionResponse{
		Messages:      []message.Message{},
		PendingEvents: []map[string]any{},
	}
	if data != nil {
		resp.Session = data.Session
		resp.Messages = data.Messages
	}

	if active := a.runs.snapshotSession(sessionID); active != nil {
		resp.ActiveRun = active.Run
		resp.Messages = active.Messages
		resp.PendingEvents = active.PendingEvents
	}

	writeJSON(w, http.StatusOK, resp)
}

func (a *app) handleDeleteSession(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("session_id")
	if a.runs.hasActiveRun(sessionID) {
		writeDetailError(w, http.StatusConflict, "session has a running task")
		return
	}
	if err := a.store.DeleteSession(sessionID); err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"status": "ok"})
}

func (a *app) handleClearSession(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("session_id")
	if a.runs.hasActiveRun(sessionID) {
		writeDetailError(w, http.StatusConflict, "session has a running task")
		return
	}
	if err := a.store.ClearSession(sessionID); err != nil {
		writeDetailError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"status": "ok"})
}
